internal/controller/workspacetemplate_controller.go

```diff
@@ -72,14 +72,21 @@ func (r *WorkspaceTemplateReconciler) Reconcile(ctx context.Context, req ctrl.Re
 
 	// Manage finalizer based on workspace usage (lazy finalizer pattern)
 	// This follows Kubernetes best practice: only add finalizers when needed
-	return r.manageFinalizer(ctx, template)
+	result, err := r.manageFinalizer(ctx, template)
+	if err != nil {
+		return result, err
+	}
+
+	return result, nil
 }
 
 // manageFinalizer implements lazy finalizer management for WorkspaceTemplates
 // Following Kubernetes best practices:
 // - Finalizers are only added when workspaces start using the template
 // - Finalizers are removed when all workspaces stop using the template
 // - This minimizes overhead and complexity compared to adding finalizers to all templates
+//
+//nolint:unparam // ctrl.Result signature maintained for consistency with controller-runtime patterns
```

Comment:
I had missed this in previous PRs, this logic won't scale:
1/ it needs to paginate
2/ it performs too many operations in a single reconciliation loop

The workspace reconciliation loop (or better: webhook) should add finalizers to the template, not the other way around.